File: wpimath/src/test/java/edu/wpi/first/math/kinematics/DifferentialDriveWheelSpeedsSerdeTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.kinematics;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import edu.wpi.first.math.proto.Kinematics.ProtobufDifferentialDriveWheelSpeeds;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import org.junit.jupiter.api.Test;

public class DifferentialDriveWheelSpeedsSerdeTest {
  private static final DifferentialDriveWheelSpeeds DATA =
      new DifferentialDriveWheelSpeeds(1.91, 2.29);
  private static final byte[] STRUCT_BUFFER =
      new byte[] {-113, -62, -11, 40, 92, -113, -2, 63, 82, -72, 30, -123, -21, 81, 2, 64};

  @Test
  void testStructPack() {
    ByteBuffer buffer = ByteBuffer.allocate(DifferentialDriveWheelSpeeds.struct.getSize());
    buffer.order(ByteOrder.LITTLE_ENDIAN);
    DifferentialDriveWheelSpeeds.struct.pack(buffer, DATA);

    byte[] actual = buffer.array();
    assertArrayEquals(actual, STRUCT_BUFFER);
  }

  @Test
  void testStructUnpack() {
    ByteBuffer buffer = ByteBuffer.wrap(STRUCT_BUFFER);
    buffer.order(ByteOrder.LITTLE_ENDIAN);

    DifferentialDriveWheelSpeeds data = DifferentialDriveWheelSpeeds.struct.unpack(buffer);
    assertEquals(DATA.leftMetersPerSecond, data.leftMetersPerSecond);
    assertEquals(DATA.rightMetersPerSecond, data.rightMetersPerSecond);
  }

  @Test
  void testProtoPack() {
    ProtobufDifferentialDriveWheelSpeeds proto = DifferentialDriveWheelSpeeds.proto.createMessage();
    DifferentialDriveWheelSpeeds.proto.pack(proto, DATA);

    assertEquals(DATA.leftMetersPerSecond, proto.getLeftMps());
    assertEquals(DATA.rightMetersPerSecond, proto.getRightMps());
  }

  @Test
  void testProtoUnpack() {
    ProtobufDifferentialDriveWheelSpeeds proto = DifferentialDriveWheelSpeeds.proto.createMessage();
    proto.setLeftMps(DATA.leftMetersPerSecond);
    proto.setRightMps(DATA.rightMetersPerSecond);

    DifferentialDriveWheelSpeeds data = DifferentialDriveWheelSpeeds.proto.unpack(proto);
    assertEquals(DATA.leftMetersPerSecond, data.leftMetersPerSecond);
    assertEquals(DATA.rightMetersPerSecond, data.rightMetersPerSecond);
  }
}

File: wpimath/src/test/native/cpp/kinematics/serde/DifferentialDriveWheelSpeedsSerdeTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <gtest/gtest.h>

#include "frc/kinematics/DifferentialDriveWheelSpeeds.h"
#include "kinematics.pb.h"

using namespace frc;

namespace {

using StructType = wpi::Struct<frc::DifferentialDriveWheelSpeeds>;
using ProtoType = wpi::Protobuf<frc::DifferentialDriveWheelSpeeds>;

constexpr std::array<uint8_t, StructType::kSize> create_test_buffer() {
  std::array<uint8_t, StructType::kSize> output;
  int buffer[] = {
    -113, -62, -11, 40, 92, -113, -2, 63, 82, -72, 30, -123, -21, 81, 2, 64
};
  for (size_t idx = 0; idx < StructType::kSize; ++idx) {
    output[idx] = static_cast<uint8_t>(buffer[idx]);
  }
  return output;
}

std::array<uint8_t, StructType::kSize> kExpectedStructBuffer =
    create_test_buffer();

constexpr DifferentialDriveWheelSpeeds kExpectedData{1.91_mps, 2.29_mps};
}  // namespace

TEST(DifferentialDriveWheelSpeedsSerdeTest, StructPack) {
  uint8_t buffer[StructType::kSize];
  StructType::Pack(buffer, kExpectedData);

  for (size_t i = 0; i < StructType::kSize; ++i) {
    EXPECT_EQ(kExpectedStructBuffer[i], buffer[i]) << " on byte " << i;
  }
}

TEST(DifferentialDriveWheelSpeedsSerdeTest, StructUnpack) {
  DifferentialDriveWheelSpeeds unpacked_data = StructType::Unpack(kExpectedStructBuffer);

    EXPECT_EQ(kExpectedData.left.value(), unpacked_data.left.value());
    EXPECT_EQ(kExpectedData.right.value(), unpacked_data.right.value());
}


TEST(DifferentialDriveWheelSpeedsSerdeTest, ProtobufPack) {
  wpi::proto::ProtobufDifferentialDriveWheelSpeeds proto;
  ProtoType::Pack(&proto, kExpectedData);

    EXPECT_EQ(kExpectedData.left.value(), proto.left_mps());
    EXPECT_EQ(kExpectedData.right.value(), proto.right_mps());
}



TEST(DifferentialDriveWheelSpeedsSerdeTest, ProtobufUnpack) {
  wpi::proto::ProtobufDifferentialDriveWheelSpeeds proto;
  proto.set_left_mps(kExpectedData.left.value());
  proto.set_right_mps(kExpectedData.right.value());

  DifferentialDriveWheelSpeeds unpacked_data = ProtoType::Unpack(proto);
    EXPECT_EQ(kExpectedData.left.value(), unpacked_data.left.value());
    EXPECT_EQ(kExpectedData.right.value(), unpacked_data.right.value());
}